src/common/transformations/src/transformations/common_optimizations/sdpa_fusion.cpp

```diff
@@ -123,31 +123,37 @@ SDPAReshapeFusion::SDPAReshapeFusion() {
     auto opt_unsq_v = optional<v1::Reshape, v0::Unsqueeze>({unsq_v, any_input()});
 
     // this Transpose may be inserted by SDPAFusionMatcher
-    auto opt_transpose_k =
-        optional<v1::Transpose>({opt_unsq_k, any_input()}, shape_matches("..., S_kv, D") && rank_more_than(2));
+    auto opt_transpose_k = optional<v1::Transpose>({opt_unsq_k, any_input()}, shape_matches("..., S_kv, D"));
 
-    auto sdpa = wrap_type<v13::ScaledDotProductAttention>({
+    auto sdpa_pattern = wrap_type<v13::ScaledDotProductAttention>({
         opt_unsq_q,
         opt_transpose_k,
         opt_unsq_v,
         any_input(),
         any_input(),
     });
 
-    auto opt_sdpa_reshape = optional<v1::Reshape, v0::Unsqueeze>({sdpa->output(0), any_input()});
-    auto opt_sdpa_transpose = optional<v1::Transpose>({opt_sdpa_reshape, any_input()});
-    auto post_sdpa = wrap_type<v1::Reshape, v0::Unsqueeze>({opt_sdpa_transpose, any_input()},
-                                                           shape_matches("Batches..., S_q, D") && rank_more_than(2));
+    auto opt_sdpa_reshape = optional<v1::Reshape, v0::Unsqueeze>({sdpa_pattern, any_input()});
+    auto post_sdpa =
+        wrap_type<v1::Reshape, v0::Unsqueeze>({opt_sdpa_reshape, any_input()}, shape_matches("Batches..., S_q, D"));
 
     ov::matcher_pass_callback callback = [OV_CAPTURE_CPY_AND_THIS](ov::pass::pattern::Matcher& m) {
         const auto& pm = m.get_pattern_value_map();
 
         auto q_node = pm.at(q);
         auto k_node = pm.at(k);
         auto v_node = pm.at(v);
-        auto sdpa_node = pm.at(sdpa).get_node_shared_ptr();
         auto post_sdpa_node = pm.at(post_sdpa).get_node_shared_ptr();
 
+        auto sdpa = ov::as_type_ptr<ov::op::v13::ScaledDotProductAttention>(pm.at(sdpa_pattern).get_node_shared_ptr());
+
+        // The mask will be ignored if causal is true; otherwise, the mask rank should be less than or equal to the SDPA
+        // input rank.
+        if (sdpa && !sdpa->get_causal() &&
+            sdpa->get_input_partial_shape(3).rank().get_length() > q_node.get_partial_shape().rank().get_length()) {
+            return false;
```

Comment:
Is it ensured at this point that the `get_input_partial_shape` will not return fully dynamic shape? Otherwise the `get_length()` can throw for dynamic rank.